wcfsetup/install/files/lib/acp/form/PackageUpdateServerAddForm.class.php

```diff
@@ -31,153 +41,167 @@ class PackageUpdateServerAddForm extends AbstractForm
+                                    [
+                                        'duplicate' => $duplicate,
+                                    ]
+                                ));
+                            }
+                        })),
+                    TextFormField::create('loginUsername')
+                        ->label('wcf.acp.updateServer.loginUsername')
+                        ->description('wcf.acp.updateServer.loginUsername.description')
+                        ->maximumLength(255),
+                    PasswordFormField::create('loginPassword')
+                        ->label('wcf.acp.updateServer.loginPassword')
+                        ->description('wcf.acp.updateServer.loginPassword.description')
+                        ->placeholder($passwordPlaceholder)
+                        ->required()
+                        ->maximumLength(255)
+                        ->addDependency(
+                            NonEmptyFormFieldDependency::create('loginUsername')
+                                ->fieldId('loginUsername')
+                        ),
+                ]),
+        ]);
     }
 
     /**
      * @inheritDoc
      */
-    public function validate()
+    public function buildForm()
     {
-        parent::validate();
+        parent::buildForm();
 
-        $this->validateServerURL();
-    }
-
-    /**
-     * Validates the server URL.
-     *
-     * @since       5.3
-     */
-    protected function validateServerURL()
-    {
-        if (empty($this->serverURL)) {
-            throw new UserInputException('serverURL');
-        }
+        $this->form->getDataHandler()->addProcessor(new CustomFormDataProcessor(
+            'password',
+            function (IFormDocument $document, array $parameters) {
+                if ($this->formAction === IFormDocument::FORM_MODE_CREATE) {
+                    return $parameters;
+                }
 
-        try {
-            $url = new Uri($this->serverURL);
-            $this->serverURL = (string)$url;
+                $username = $parameters['data']['loginUsername'] ?? null;
+                \assert($this->formObject instanceof PackageUpdateServer);
 
-            if (!$url->getHost()) {
-                throw new UserInputException('serverURL', 'invalid');
-            }
-            if ($url->getHost() !== 'localhost') {
-                if ($url->getScheme() !== 'https') {
-                    throw new UserInputException('serverURL', 'invalidScheme');
-                }
-                if ($url->getPort()) {
-                    throw new UserInputException('serverURL', 'nonStandardPort');
+                if ($username === '') {
+                    $parameters['data']['loginUsername'] = '';
+                    $parameters['data']['loginPassword'] = '';
+                } elseif (
+                    $username === $this->formObject->loginUsername && empty($parameters['data']['loginPassword'])
```

Comment:
```suggestion
                    $username === $this->formObject->loginUsername
                    && $parameters['data']['loginPassword'] === ''
```
If you already please the condition into multiple lines, then also split at the binary boolean operators. Also: `empty`.